File: ChainOfResponsibility/ChainOfResponsibilityShop/ChainOfResponsibilityShop/BaseDiscountHandler.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ChainOfResponsibilityShop
{
    public abstract class BaseDiscountHandler : IDiscountHandler
    {
        private IDiscountHandler? _next;

        public void SetNext(IDiscountHandler next)
        {
            _next = next;
        }

        protected void Next(string clientType)
        {
            if (_next != null)
            {
                _next.GetDiscount(clientType);
            }
            else
            {
                Console.WriteLine("No discount available for this client type.");
            }
        }

        public abstract void GetDiscount(string clientType);
        
    }
}

File: ChainOfResponsibility/JavaExample/ChainOfResponsibility/src/BaseDiscountHandler.java
public abstract class BaseDiscountHandler implements IDiscountHandler{
    private IDiscountHandler next;

    @Override
    public void setNext(IDiscountHandler next) {
        this.next = next;
    }

    protected void next(String clientType) {
        if (this.next != null) {
            this.next.getDiscount(clientType);
        } else {
            System.out.println("No discount available for this client type.");
        }
    }

    public abstract void getDiscount(String clientType);
}
